logic/Rule.java

```diff
@@ -402,12 +402,16 @@ public void validate(LogicManager logicMgr, ConceptManager conceptMgr) {
         }
 
         private void validateInsertable(LogicManager logicMgr) {
-            FunctionalIterator<Map<Identifier.Variable.Name, Label>> whenCombinations = logicMgr.typeInference().typePermutations(rule.when, false);
-            Set<Map<Identifier.Variable.Name, Label>> allowedThenCombinations = logicMgr.typeInference().typePermutations(rule.then, true).toSet();
+            Set<Identifier.Variable.Retrievable> sharedIDs = iterate(rule.then.identifiers())
+                    .filter(thenID -> thenID.isName() && rule.when.identifiers().contains(thenID))
+                    .map(Identifier.Variable::asRetrievable).toSet();
+            FunctionalIterator<Map<Identifier.Variable.Name, Label>> whenCombinations = logicMgr.typeInference().typePermutations(rule.when, false, sharedIDs);
+            Set<Map<Identifier.Variable.Name, Label>> insertableThenCombinations = logicMgr.typeInference().typePermutations(rule.then, true, sharedIDs).toSet();
 
             whenCombinations.forEachRemaining(nameLabelMap -> {
-                if (allowedThenCombinations.stream().noneMatch(thenMap -> nameLabelMap.entrySet().containsAll(thenMap.entrySet())))
+                if (!insertableThenCombinations.contains(nameLabelMap)) {
```

Comment:
using a `contains` instead of a `stream()` converts a O(n) operation into an O(1) operation!